x/bank/keeper/keeper.go

```diff
@@ -398,28 +372,36 @@ func (k BaseKeeper) BurnCoins(ctx sdk.Context, moduleName string, amt sdk.Coins)
 		panic(sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "module account %s does not have permissions to burn tokens", moduleName))
 	}
 
-	err := k.subUnlockedCoins(ctx, acc.GetAddress(), amt)
+	err := k.subUnlockedCoins(ctx, acc.GetAddress(), amounts)
 	if err != nil {
 		return err
 	}
 
-	// update total supply
-	supply := k.GetTotalSupply(ctx)
-	supply = supply.Sub(amt)
-
-	k.setSupply(ctx, supply)
+	for _, amount := range amounts {
+		supply := k.GetSupply(ctx, amount.GetDenom())
+		supply = supply.Sub(amount)
+		k.setSupply(ctx, supply)
+	}
 
 	logger := k.Logger(ctx)
-	logger.Info("burned tokens from module account", "amount", amt.String(), "from", moduleName)
+	logger.Info("burned tokens from module account", "amount", amounts.String(), "from", moduleName)
 
 	// emit burn event
 	ctx.EventManager().EmitEvent(
-		types.NewCoinBurnEvent(acc.GetAddress(), amt),
+		types.NewCoinBurnEvent(acc.GetAddress(), amounts),
 	)
 
 	return nil
 }
 
+func (k BaseKeeper) setSupply(ctx sdk.Context, coin sdk.Coin) {
+	store := ctx.KVStore(k.storeKey)
+	supplyStore := prefix.NewStore(store, types.SupplyKey)
+
+	bz := k.cdc.MustMarshalBinaryBare(&coin)
+	supplyStore.Set([]byte(coin.GetDenom()), bz)
+}
```

Comment:
Why not keeping a loop here?
1. in all places where we use this function we have the following pattern: https://github.com/cosmos/cosmos-sdk/pull/8950/files#diff-9994bff9a7f71878caa036e9c924ab8b06e31490a17874fab72648ed3404d220R346
2. if we move the loop here, we don't need to recreate and `store` and do additional allocations. @odeke-em was shaving recently such places.

Reply:
An argument could be made for doing all of this logic inline inside `MintCoins` and `BurnCoins`

Reply:
@jgimeno I wonder if there's value in making this `setSupply(sdk.KVStore, sdk.Coin)` to avoid the allocations from `ctx.KVStore` as @robert-zaremba is suggesting. Wdyt?